syntax = "proto3";

message User {
    string userID = 1;
}

message Post {
    string postID = 1;
    string title = 2;
    string text = 3;
    string mediaUrl = 4;
    string score = 5;
    string state = 6;
    string publicationDate = 7;
    User author = 8;
    String post_created_message = 9;
    string upVote = 6;
    string downVote = 7;
}

message Comment {
    string commentID = 1;
    User author = 2;
    string score = 3;
    string state = 4;
    string publicationDate = 5;
    string upVote = 6;
    string downVote = 7;
}

message N {
    int number; 
}

service Reddit {
    rpc CreatePost(Post) returns (Post) {}
    rpc UpVote(Post) returns (Post) {}
    rpc DownVote(Post) returns (Post) {}
    rpc Retrieve(Post) returns (Post) {}
    rpc CreateComment(Comment) returns (Comment) {}
    rpc RetrieveListOfNMostUpvotedComments(Post,N) returns (stream Comment) {}
    rpc ExpandCommentBranch(Comment) returns (Comment) {}
}